### PR TITLE
adds new CPU kernel for SGD op supporting BF16 data type

### PR DESCRIPTION
**Pull request by arogowie-intel**
### PR types
 New features

### PR changes
OPs 

### Describe
This PR adds new CPU kernel for SGD op supporting BF16 data type.

### CODE REVIEW

**Comment by arogowie-intel** [comment]
@wozna @arlesniak @jczaja Could you start review please?

File: paddle/fluid/operators/optimizers/sgd_op.h [hunk 1]
```diff
@@ -179,5 +180,113 @@ class SGDOpKernel<platform::CPUDeviceContext, T>
...
+
+        // for distributed training, a sparse var may be empty,
+        // just skip updating.
+        if (grad->rows().size() == 0) {
+          return;
+        }
+
+        auto grad_height = grad->height();
+        auto out_dims = param_out->dims();
+        PADDLE_ENFORCE_EQ(grad_height, out_dims[0]);
```
Comment (wozna) [inline paddle/fluid/operators/optimizers/sgd_op.h]: Please add type and comment for this enforcement.

File: paddle/fluid/operators/optimizers/sgd_op.h [hunk 2]
```diff
@@ -179,5 +180,113 @@ class SGDOpKernel<platform::CPUDeviceContext, T>
...
+
+        size_t grad_row_numel = grad_value.numel() / grad_rows.size();
+        PADDLE_ENFORCE_EQ(static_cast<int64_t>(grad_row_numel),
+                          param_out->numel() / grad_height);
+
+        auto *grad_data = grad_value.data<T>();
+        auto *out_data = param_out->data<T>();
+        auto *lr = learning_rate->data<T>();
+        for (size_t i = 0; i < grad_rows.size(); i++) {
+          PADDLE_ENFORCE(grad_rows[i] < grad_height,
```
Comment (wozna) [inline paddle/fluid/operators/optimizers/sgd_op.h]: We should use more specific PADDLE_ENFORCE like PADDLE_ENFORCE_LT. Everything mentioned on the website mentioned above.

File: paddle/fluid/operators/optimizers/sgd_op.h [hunk 3]
```diff
@@ -179,5 +180,113 @@ class SGDOpKernel<platform::CPUDeviceContext, T>
...
+          for (size_t j = 0; j < grad_row_numel; j++) {
+            out_data[grad_rows[i] * grad_row_numel + j] -=
+                lr[0] * grad_data[i * grad_row_numel + j];
+          }
+        }
+      } else {
+        PADDLE_THROW("Unsupported Variable Type of Grad");
+      }
+    } else if (param_var->IsType<framework::SelectedRows>()) {
+      PADDLE_ENFORCE(grad_var->IsType<framework::SelectedRows>(),
```
Comment (wozna) [inline paddle/fluid/operators/optimizers/sgd_op.h]: Ditto

File: paddle/fluid/operators/optimizers/sgd_op.h [hunk 4]
```diff
@@ -179,5 +180,113 @@ class SGDOpKernel<platform::CPUDeviceContext, T>
...
+        return;
+      }
+
+      auto param_row_width = param.value().dims()[1];
+      auto grad_row_width = grad.value().dims()[1];
+      VLOG(4) << " param rows: " << param.rows().size()
+              << " param memory rows: " << param.value().dims()[0]
+              << " grad rows: " << grad.rows().size()
+              << " grad memory rows: " << grad.value().dims()[0];
+      PADDLE_ENFORCE_EQ(param_row_width, grad_row_width,
```
Comment (wozna) [inline paddle/fluid/operators/optimizers/sgd_op.h]: Ditto

File: paddle/fluid/operators/optimizers/sgd_op.h [hunk 5]
```diff
@@ -179,5 +180,113 @@ class SGDOpKernel<platform::CPUDeviceContext, T>
...
+                        "param_row should have the same size with grad_row");
+
+      const auto *lr = learning_rate->data<T>();
+      const auto *grad_data = grad.value().data<T>();
+      auto *out_data = param_out->mutable_value()->data<T>();
+      for (size_t i = 0; i < grad.rows().size(); i++) {
+        int64_t id_index = param_out->AutoGrownIndex(grad.rows()[i], false);
+        PADDLE_ENFORCE_GE(id_index, static_cast<int64_t>(0),
```
Comment (wozna) [inline paddle/fluid/operators/optimizers/sgd_op.h]: Ditto

**Review by wozna — COMMENTED** [review]
Everything I mentioned is about PADDLE_ENFORCE 
On this site, there is information on how we should write error messages: https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification-(English-Verison)

**Comment by arlesniak** [comment]
Please make integration test i.e. **test_fit_a_line.py** to verify this implementation (it can be done in next PR)

File: paddle/fluid/operators/optimizers/sgd_op.h (hunk 1 above)
Comment (arogowie-intel) [reply paddle/fluid/operators/optimizers/sgd_op.h]: Done

File: paddle/fluid/operators/optimizers/sgd_op.h (hunk 2 above)
Comment (arogowie-intel) [reply paddle/fluid/operators/optimizers/sgd_op.h]: Done

File: paddle/fluid/operators/optimizers/sgd_op.h (hunk 4 above)
Comment (arogowie-intel) [reply paddle/fluid/operators/optimizers/sgd_op.h]: Done

File: paddle/fluid/operators/optimizers/sgd_op.h (hunk 3 above)
Comment (arogowie-intel) [reply paddle/fluid/operators/optimizers/sgd_op.h]: Done

**Comment by arogowie-intel** [comment]
@wozna, @arlesniak Please review again.

**Comment by arogowie-intel** [comment]
@luotao1 
The PR-CI-APPROVAL fails because of:
1.
> 0. Unittest is not allowed to be disabled.

I'm using
```
@unittest.skipIf(not core.supports_bfloat16(),
```
To not run tests on platforms which does not support bfloat16 data type.

2.
>  1. Developers are not allowed to set the check_dygraph field directly, which is set to True by default.

I'm turning off dygraph mode since we does not support it with bfloat16 yet.

**Comment by arogowie-intel** [comment]
@luotao1 Could you help with PR-CI-ROCM-Compile?
It somehow fails during compilation, however locally I do not see such an error. Moreover no other CI run failed with such an error.

**Comment by luotao1** [comment]
@qili93 Could you help see error of PR-CI-ROCM-Compile?

**Comment by luotao1** [comment]
@arogowie-intel You can see the ROCM build step in https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/rocm_docs/paddle_install_cn.html

**Comment by qili93** [comment]
Hi @arogowie-intel to fix the build error or "PR-CI-ROCM-Compile", please change your code as following:

```c++
 PADDLE_ENFORCE(grad_var->IsType<framework::SelectedRows>(),
# ------ change the above code to below ----------
 PADDLE_ENFORCE_EQ(grad_var->IsType<framework::SelectedRows>(), true, 
```

ROCM defined `PADDLE_ENFORCE` as a device function in `enforce.h`, so please avoid to use it in host functions.


```log
2021-04-12 21:28:13 In file included from /paddle/paddle/fluid/operators/optimizers/sgd_op.cu:16:
2021-04-12 21:28:13 /paddle/paddle/fluid/operators/optimizers/sgd_op.h:250:22: error: expected ')'
2021-04-12 21:28:13                      platform::errors::InvalidArgument(
2021-04-12 21:28:13                      ^
2021-04-12 21:28:13 /paddle/paddle/fluid/operators/optimizers/sgd_op.h:249:7: note: to match this '('
2021-04-12 21:28:13       PADDLE_ENFORCE(grad_var->IsType<framework::SelectedRows>(),
2021-04-12 21:28:13       ^
2021-04-12 21:28:13 /paddle/paddle/fluid/platform/enforce.h:438:13: note: expanded from macro 'PADDLE_ENFORCE'
2021-04-12 21:28:13       printf("Error: %s:%d Assertion `%s` failed. " __FORMAT "\n", __FILE__, \
2021-04-12 21:28:13             ^
```

**Review by arlesniak — APPROVED** [review]
LGTM

**Comment by arogowie-intel** [comment]
@luotao1  Could you please start your review?

**Comment by arogowie-intel** [comment]
@luotao1 I'm waiting for the last CI `pm-test` run, which is in "_Pending_" status from yesterday. Moreover it is not required one. Regarding the PR-CI-APPROVAL I have explained why it is failing in my earlier comment. 
Could you start your review please? This PR is quite important for continuation of word2vec training work.

**Review by luotao1 — APPROVED** [review]

**Comment by luotao1** [comment]
> I'm waiting for the last CI pm-test run, which is in "Pending" status from yesterday. Moreover it is not required one.

@arogowie-intel You can ignore the CI which is not the required one.

**Pull request merged**